packages/medusa/src/services/payment-collection.ts

```diff
@@ -160,4 +253,479 @@ export default class PaymentCollectionService extends TransactionBaseService {
+        if (sess.id === sessionId) {
+          return sessionRefreshed
+        }
+        return sess
+      })
+
+      if (session.payment_authorized_at) {
+        payCol.authorized_amount -= session.amount
+      }
+
+      const paymentCollectionRepository = manager.getCustomRepository(
+        this.paymentCollectionRepository_
+      )
+
+      await paymentCollectionRepository.save(payCol)
+
+      return sessionRefreshed
+    })
+  }
+
+  async authorize(
+    paymentCollectionId: string,
+    context: Record<string, unknown> = {}
+  ): Promise<PaymentCollection> {
+    return await this.atomicPhase_(async (manager: EntityManager) => {
+      const paymentCollectionRepository = manager.getCustomRepository(
+        this.paymentCollectionRepository_
+      )
+
+      const payCol = await this.retrieve(paymentCollectionId, {
+        relations: ["payment_sessions", "payments"],
+      })
+
+      if (payCol.authorized_amount === payCol.amount) {
+        return payCol
+      }
+
+      // If cart total is 0, we don't perform anything payment related
+      if (payCol.amount <= 0) {
+        payCol.authorized_amount = 0
+        return await paymentCollectionRepository.save(payCol)
+      }
+
+      if (!payCol.payment_sessions?.length) {
+        throw new MedusaError(
+          MedusaError.Types.NOT_ALLOWED,
+          "You cannot complete a Payment Collection without a payment session."
+        )
+      }
+
+      let authorizedAmount = 0
+      for (const session of payCol.payment_sessions) {
+        if (session.payment_authorized_at) {
+          authorizedAmount += session.amount
+          continue
+        }
+
+        const auth = await this.paymentProviderService_
+          .withTransaction(manager)
+          .authorizePayment(session, context)
+
+        if (auth?.status === PaymentSessionStatus.AUTHORIZED) {
+          authorizedAmount += session.amount
+
+          const inputData: Omit<PaymentProviderDataInput, "customer"> = {
+            amount: session.amount,
+            currency_code: payCol.currency_code,
+            provider_id: session.provider_id,
+            resource_id: payCol.id,
+          }
+
+          payCol.payments.push(
+            await this.paymentProviderService_
+              .withTransaction(manager)
+              .createPaymentNew(inputData)
+          )
+        }
+      }
+
+      if (authorizedAmount === 0) {
+        payCol.status = PaymentCollectionStatus.AWAITING
+      } else if (authorizedAmount < payCol.amount) {
+        payCol.status = PaymentCollectionStatus.PARTIALLY_AUTHORIZED
+      } else if (authorizedAmount === payCol.amount) {
+        payCol.status = PaymentCollectionStatus.AUTHORIZED
+      }
+
+      payCol.authorized_amount = authorizedAmount
+      const payColCopy = await paymentCollectionRepository.save(payCol)
+
+      await this.eventBusService_
+        .withTransaction(manager)
+        .emit(PaymentCollectionService.Events.UPDATED, payColCopy)
```

Comment:
question: Would it make sense to make an event specific to the authorization of payment collections as I feel `updated` might be a bit too generic?

Reply:
Yeah, that makes sense.